Introducing new storage models required for Contributor Admin Dashboard: GSoC M1.2

kshitij01042002 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR introduces storage model changes required for #17440 
2. This PR does the following: Added four new storage models to store the stats required to be displayed in contributor admin dashboard. 
fetch_page is a pagination function implemented so that we can use it in future while building frontend. It supports pagination, sorting and filtering. Test cases have been added for each unique condition.
The reasoning for introducing the models and more information about the models can be found at [GSoC Proposal (2023) -- 2.1 Contributor Dashboard Admin Stats Table (Kshitij Patil)](https://docs.google.com/document/d/19lCEYQUgV7_DwIK_0rz3zslRHX2qKOHn-t9Twpi0qu0/edit)

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x]  The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

Proof that all the test cases are covered

![Screenshot from 2023-06-05 19-40-45](https://github.com/oppia/oppia/assets/91470808/ef9bce58-e09e-4d5a-bb16-d3c3c50d7b8a)


<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

kshitij01042002 committed:
failing tests

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
restored changed file

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
initial commit

kshitij01042002 committed:
TranslationSubmitterTotalContributionStatsModel added

kshitij01042002 committed:
Added QuestionSubmitterTotalContributionStatsModel and wipeout service tests

kshitij01042002 committed:
Added TranslationReviewerTotalContributionStatsModel

kshitij01042002 committed:
added QuestionReviewerTotalContributionStatsModel

kshitij01042002 committed:
tests for new storage models

kshitij01042002 committed:
Added pagination features and test cases

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.2

seanlip commented on assets/constants.ts:
```diff
@@ -115,6 +115,22 @@ export default {
+    "DecreasingPerformance": "Highest to Lowest Performance",
+    "IncreasingAccuracy": "Lowest to Highest Accuracy",
+    "DecreasingAccuracy": "Highest to Lowest Accuracy",
+    "IncreasingSubmissions": "Lowest to Highest submitted Translations",
```
Perhaps you should capitalize Submitted and Reviewed (in 4 places below) as well.

kshitij01042002 replied:
Changed it in the new commits.

seanlip commented on core/domain/takeout_service_test.py:
```diff
@@ -764,6 +778,90 @@ def set_up_non_trivial(self) -> None:
+                self.LAST_HUNDRED_ACCEPTED_TRANSLATIONS_COUNT),
+            last_hundred_accepted_translations_without_reviewer_edits_count=(
+                self
+            # Pylint disable is needed cause variable name is too long.
```
cause --> because

kshitij01042002 replied:
Done.

seanlip commented on core/domain/takeout_service_test.py:
```diff
@@ -1873,6 +1991,113 @@ def test_export_data_for_full_user_nontrivial_is_correct(self) -> None:
+                    'overall_accuracy': self.OVERALL_ACCURACY,
+                    'last_hundred_accepted_translations_count': (
+                        self.LAST_HUNDRED_ACCEPTED_TRANSLATIONS_COUNT),
+                    # Pylint disable is needed cause variable name is too long.
```
cause --> because

kshitij01042002 replied:
Done.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +1997,1360 @@ def export_data(
+    # last 100 reviews.
+    last_hundred_accepted_translations_without_reviewer_edits_count = (
+        datastore_services.IntegerProperty(required=True, indexed=True))
+    # The number of rejected translation in last 100 reviews.
```
translation --> translations

kshitij01042002 replied:
Done.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +1997,1360 @@ def export_data(
+                    after this batch.
+        """
+        sort = -cls.recent_performance
+        if sort_by == (
```
This seems indirect, can't you just pass 'IncreasingLastActivity' etc. as the sort_by ? You should use IDs for these things rather than displayable names, since the former are stable and the latter are changeable.

Then I would also suggest just making a dict mapping those IDs to the corresponding sort property and doing a dict lookup.

kshitij01042002 replied:
Added a new dict **sort_options_dict** . In future if we want to add or remove any sort propoerty we just have to modify that dict rather than tweaking the whole if-else ladder. Thanks for the suggestion.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +1997,1360 @@ def export_data(
+                datastore_services.all_of(
+                    cls.language_code == language_code,
+                    cls.topic_ids_with_translation_submissions.IN(topic_ids)
+                    )).order(sort)
```
Deindent by 4 to match opening brace.

kshitij01042002 replied:
Done,

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +1997,1360 @@ def export_data(
+        else:
+            sort_query = cls.query(
+                datastore_services.all_of(
+                    cls.language_code == language_code
+                    )).order(sort)
```
Ditto.

kshitij01042002 replied:
Done.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +1997,1360 @@ def export_data(
+        elif sort_by == (
+                constants.CD_ADMIN_STATS_SORT_OPTIONS[
+                    'IncreasingSubmissions']):
+            sort = cls.submitted_translations_count
```
else raise Exception?

kshitij01042002 replied:
If None is passed, it will be set to -cls.recent_performance as default, as you can see on line 2203.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +1997,1360 @@ def export_data(
+
+        sorted_results: List[
+            TranslationSubmitterTotalContributionStatsModel] = []
+        num_models_per_fetch = 100
```
Make this a named constant (or use an existing named constant), declare it at the top of the file, and use that constant throughout all the methods which need it.

kshitij01042002 replied:
Made a new constant in the file NUM_MODELS_PER_FETCH

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +1997,1360 @@ def export_data(
+                        if len(sorted_results) == page_size:
+                            break
+        else:
+            sorted_results = list((
```
You can drop one of the parens at the end (and the corresponding close paren).

kshitij01042002 replied:
Done.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +1997,1360 @@ def export_data(
+        num_models_per_fetch = 100
+        today = datetime.date.today()
+
+        if last_activity is not None:
```
You can't use a filter for this in the datastore query above?

kshitij01042002 replied:
We can't because of the property of gae models that states: 
**The first sort property must be the same as the property to which an inequality filter is applied.** 
As we are sorting on multiple metrics we have to go by this method to apply filter for last_activity.

I have mentioned this on line 2233 as well, where we are writing sort query.

seanlip replied:
I think you can make the comment in line 2233 more clear. Be specific about where the restriction comes from, provide a link, give the reader enough context on what you're trying to do and why it can't be done. Assume your reader has no knowledge of GAE query restrictions or this conversation.

It's fine for the comment to be multiple paragraphs long, as long as it's clear to any developer who reads it.

kshitij01042002 replied:
I have added a reference link to gae docs stating that error, where a new developer can easily understand why we didn't used the inequality filter with the sort query.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +1997,1360 @@ def export_data(
+            cls.get_all().filter(cls.contributor_id == user_id).fetch())
+        for model in stats_models:
+            splitted_id = model.id.split('.')
+            id_without_user_id = '%s' % (splitted_id[0])
```
Why not just `splitted_id[0]` (without the %s stuff)?

kshitij01042002 replied:
Yeah, that sounds fine. Changed.

seanlip reviewed (changes requested):
Thanks @kshitij01042002 -- I left comments for only one of the models but similar comments apply throughout. PTAL.

kshitij01042002 committed:
changes requested by reviewers

kshitij01042002 committed:
Merge branch 'gsoc_M1.2' of https://github.com/kshitij01042002/oppia into gsoc_M1.2

kshitij01042002 committed:
fix lint errors

kshitij01042002 commented:
Hi @seanlip I have pushed new commits according to your suggestions and also answered other comments. PTAL.

Thanks for the review.

kshitij01042002 committed:
fixed typo

seanlip reviewed (changes requested):
@kshitij01042002 Per our [PR process](https://github.com/oppia/oppia/wiki/Make-a-pull-request), please reply to and address all comments fully before requesting subsequent reviews. I'll wait to review until you've done that.

kshitij01042002 committed:
added link to gae documentation

kshitij01042002 commented:
Hi @seanlip addressed all the comments. PTAL. Thanks.

kshitij01042002 commented:
Making this PR as draft since some new changes might be needed.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.2

kshitij01042002 committed:
made new changes suggested by mentor and reviewer

kshitij01042002 commented:
Hi @seanlip I have addressed all the comments. PTAL. thanks.

lkbhitesh07 commented on assets/constants.ts:
```diff
@@ -115,6 +115,22 @@ export default {
+    "DecreasingAccuracy": "Highest to Lowest Accuracy",
+    "IncreasingSubmissions": "Lowest to Highest Submitted Contributions",
+    "DecreasingSubmissions": "Highest to Lowest Submitted Contributions",
+    "IncreasingReviewedTranslation": "Lowest to Highest Reviewed Translations",
+    "DecreasingReviewedTranslation": "Highest to Lowest Reviewed Translations",
+    "IncreasingReviewedQuestion": "Lowest to Highest Reviewed Questions",
+    "DecreasingReviewedQuestion": "Highest to Lowest Reviewed Questions"
```
```suggestion
    "IncreasingReviewedTranslations": "Lowest to Highest Reviewed Translations",
    "DecreasingReviewedTranslations": "Highest to Lowest Reviewed Translations",
    "IncreasingReviewedQuestions": "Lowest to Highest Reviewed Questions",
    "DecreasingReviewedQuestions": "Highest to Lowest Reviewed Questions"
```

kshitij01042002 replied:
Changed it and also the references to these changes. Thanks.

lkbhitesh07 commented on assets/constants.ts:
```diff
@@ -6977,6 +6993,6 @@ export default {
   ],
   "CONTRIBUTOR_CERTIFICATE_WIDTH": 1493,
   "CONTRIBUTOR_CERTIFICATE_HEIGHT": 1313,
-  "BRANCH_NAME": "",
-  "SHORT_COMMIT_HASH": ""
+  "BRANCH_NAME": "gsoc_M1.2",
+  "SHORT_COMMIT_HASH": "e526997625"
```
Revert this.

kshitij01042002 replied:
Actually I didn't made this change. Tho I have reverted it, I am not sure how it was changed.

lkbhitesh07 reviewed (changes requested):
Thanks for the changes @kshitij01042002. Reviewed the codeowner file and left 2 comments, please take a look.
Thanks

kshitij01042002 committed:
changes requested by hitesh

kshitij01042002 commented:
Thanks @lkbhitesh07 for the review. I have addressed your comments. PTAL.

lkbhitesh07 reviewed (approved):
Thanks for the changes @kshitij01042002. LGTM for the codeowner file.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -43,13 +43,20 @@
 STATUS_ACCEPTED: Final = 'accepted'
 STATUS_IN_REVIEW: Final = 'review'
 STATUS_REJECTED: Final = 'rejected'
+STATUS_ACCEPTED_WITH_EDITS: Final = 'accepted_with_edits'
```
I advise that you don't call these statuses. This is going to confuse things. Instead you should use REVIEW_OUTCOME_ACCEPTED, REVIEW_OUTCOME_ACCEPTED_WITH_EDITS, etc. 

Otherwise, with what you have, it leads to questions about why STATUS_CHOICES doesn't include STATUS_ACCEPTED_WITH_EDITS etc.

kshitij01042002 replied:
Changed.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -43,13 +43,20 @@
     STATUS_REJECTED
 ]
 
+REVIEW_OUTCOME_STATUS: Final = [
```
REVIEW_OUTCOME_CHOICES

kshitij01042002 replied:
Updated.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+    """Records the Total Translation contribution stats and data of
+    recent_review keyed per (contributor_id, language_code) tuple.
+    Its IDs will be in the following
+        structure: [contributor_id][language_code]
```
This is not quite right, it doesn't match your implementation.

`The ID of each model is formatted as [language_code].[contributor_id]`

Please make sure other docstrings are fully correct as well, otherwise we go through unnecessary review cycles.

kshitij01042002 replied:
Updated

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+    # The user ID of the translation contributor.
+    contributor_id = datastore_services.StringProperty(
+        required=True, indexed=True)
+    # The topic ID(s) of the topics the contributor has atleast one
```
...of the topics for which the contributor has at least one ...

Note also that "at least" is two words. Please fix here and elsewhere.

kshitij01042002 replied:
Fixed.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+    recent_performance = datastore_services.IntegerProperty(
+        required=True, indexed=True)
+    # Overall accuracy of the user.
+    # overall_accuracy = accepted cards/ submitted cards.
```
Add a space before /

kshitij01042002 replied:
Added.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+        last_contribution_date: datetime.date
+    ) -> str:
+        """Creates a new TranslationSubmitterTotalContributionStatsModel
+        instance and returns its ID.
```
Missing args and returns in docstring. Please fix here and elsewhere.

kshitij01042002 replied:
Added docstring.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+        cls, language_code: str, contributor_id: str
+    ) -> Optional[TranslationSubmitterTotalContributionStatsModel]:
+        """Gets the TranslationSubmitterTotalContributionStatsModel
+        matching the supplied language_code, contributor_id, topic_id.
```
This docstring is wrong. Please proofread your docstrings.

kshitij01042002 replied:
Fixed.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+            such model instance exists.
+        """
+        entity_id = cls.construct_id(
+            language_code, contributor_id)
```
Nit: can be on the previous line.

kshitij01042002 replied:
Changed.

seanlip replied:
Doesn't seem changed to me.

kshitij01042002 replied:
Sorry, missed it. Changed this as well as other with similarities.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+            page_size: int. Number of models to fetch.
+            offset: int. Number of results to skip from the beginning of all
+                results matching the query.
+            sort_by: str|None. A string indicating how to sort the result.
```
Specify the possible choices that are allowed, and use a more restrictive Python type that limits to those choices.

kshitij01042002 replied:
Made new class **SortChoice** to act as restriction for sort_by

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+                results matching the query.
+            sort_by: str|None. A string indicating how to sort the result.
+            topic_ids: List[str]|None.
+                List of topics user has contributions.
```
You can move the start of this sentence to the end of the previous line.

Also I don't think this is correct? Isn't this the thing that the caller is passing in rather than some attribute of the user? I.e. "List of topic IDs to fetch the contribution stats for." or something like that? There is no specific user here, I think, so I'm not sure what your docstring is referring to.

kshitij01042002 replied:
Yeah you are correct, it is not related to user but passed down from frontend to fetch contributor stats.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+            sort_by: str|None. A string indicating how to sort the result.
+            topic_ids: List[str]|None.
+                List of topics user has contributions.
+            last_activity: int. Number of days within which a user has
```
I don't understand what this means. It might have the same problem as the previous one.

Also the variable name doesn't sound like an int. Perhaps num_days_since_last_activity or similar.

kshitij01042002 replied:
Updated. PTAL.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+        sort_by: Optional[str],
+        topic_ids: Optional[List[str]],
+        last_activity: Optional[int],
+        language_code: str
```
The optional args should go at the end.

kshitij01042002 replied:
Done. Will keep in mind from next time.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+                List of topics user has contributions.
+            last_activity: int. Number of days within which a user has
+                contributed.
+            language_code: str. Language Code for filter.
```
The language code to get results for.

kshitij01042002 replied:
Updated.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+            3-tuple(sorted_results, next_offset, more). where:
+                sorted_results:
+                    list(TranslationSubmitterTotalContributionStatsModel).
+                    The list of models with supplied filters and sort_by.
```
The list of models which match the supplied language_code, topic_ids and last_activity filters, returned in the order specified by sort_by.

kshitij01042002 replied:
Updated.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+        """
+
+        sort_options_dict = {
+            constants.CD_ADMIN_STATS_SORT_OPTIONS['IncreasingLastActivity']:
```
The string value appear to be duplicated, so you should make them constants and put them at the top, e.g. SORT_KEY_INCREASING_LAST_ACTIVITY.

Also, as mentioned in a previous comment, the mapping should be directly from SORT_KEY_INCREASING_LAST_ACTIVITY and you should be passing the sort keys to this function, rather than the human-readable values.

kshitij01042002 replied:
Updated. PTAL.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+        # The first sort property must be the same as the property to which
+        # an inequality filter is applied. Thus, the inequality filter on
+        # last_activity can not be used here. Learn more about this here
+        # https://cloud.google.com/appengine/docs/legacy/standard/go111/datastore/query-restrictions#properties_used_in_inequality_filters_must_be_sorted_first.
```
Add a conclusion at the end: This is why we are handling the last_activity inequality filter separately below.

kshitij01042002 replied:
I added the reasoning and conclusion before reference link. PTAL

seanlip replied:
"below" is ambiguous since it can refer to what's just "below" this comment. This is why my original suggestion included "separately".

kshitij01042002 replied:
Changed. PTAL.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+                result_models: Sequence[
+                    TranslationSubmitterTotalContributionStatsModel] = (
+                    sort_query.fetch(
+                        NUM_MODELS_PER_FETCH, offset=next_offset))
```
I think you can make NUM_MODELS_PER_FETCH larger, say 500. Otherwise you'll potentially end up fetching quite a lot of pages in the case where you'll need to retrieve all the results. It's better to have fewer queries.

kshitij01042002 replied:
Yeah sg. Changed it to 500.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+        # last_activity can not be used here. Learn more about this here
+        # https://cloud.google.com/appengine/docs/legacy/standard/go111/datastore/query-restrictions#properties_used_in_inequality_filters_must_be_sorted_first.
+        if topic_ids is not None:
+            sort_query = cls.query(
```
You will need to exercise this manually on your local server to generate the index.yaml file indexes needed to serve this query. https://cloud.google.com/appengine/docs/legacy/standard/python/config/indexref#creating_index_files

kshitij01042002 replied:
Hi @seanlip need a little help here. How can I run the queries manually on the local server. I am bit confused here. The documentation you linked suggests to run the queries which will create a index.yaml file automatically. Should I manually type-in the values?

seanlip replied:
https://github.com/oppia/oppia/wiki/Debugging-datastore-locally might be useful. You can make queries from there.

kshitij01042002 replied:
I am trying this might take some time, I figuring out something that will help me through out the project and not just for this case.

kshitij01042002 replied:
Thanks. That helps.

kshitij01042002 replied:
Hi @seanlip I tried running the queries manually but every index was not generated in index.yaml. I have added the indexes with all possible queries in our models. PTAL.

seanlip replied:
That's a bit odd, usually we have the opposite problem (i.e. people seeing unexpected changes in index.yaml). Maybe that will trigger when you run the devserver once you've hooked everything up, but for now it's also fine to populate them manually. Thanks.

kshitij01042002 replied:
Actually I tested it on the devserver and ran queries, some indexes were generated for that matter but added the missed ones.

seanlip commented on assets/constants.ts:
```diff
@@ -115,6 +115,22 @@ export default {
   },
+
+  // Sorting categories for contributor admin dashboard.
+  "CD_ADMIN_STATS_SORT_OPTIONS": {
```
You should have a backend test that verifies that the keys here match the constants defined in the backend file (see comment below). This will help catch errors if a developer changes the key in the future.

kshitij01042002 replied:
Made the sort_choices more strict below PTAL.

seanlip replied:
Could you please point me to the backend test? There should still be one to ensure that the constants defined here and the ones defined in the backend don't diverge.

kshitij01042002 replied:
Where should this backend test go?

seanlip replied:
I would put it in the test file associated with the backend file which defines those constants. In this case they would go in core/storage/suggestions/gae_models_test.py.

kshitij01042002 replied:
Added it there. PTAL.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+                    sort_query.fetch(
+                        NUM_MODELS_PER_FETCH, offset=next_offset))
+                if not result_models:
+                    next_offset += 1
```
Why are you incrementing in this case?

kshitij01042002 replied:
It shouldn't be incrementing here. Thanks

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+        last_contribution_date: datetime.date
+    ) -> str:
+        """Creates a new TranslationReviewerTotalContributionStatsModel
+        instance and returns its ID.
```
Missing args and returns.

kshitij01042002 replied:
Added.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+            page_size: int. Number of models to fetch.
+            offset: int. Number of results to skip from the beginning of all
+                results matching the query.
+            sort_by: str|None. A string indicating how to sort the result.
```
Same concerns as above, here and for other parts of this method.

kshitij01042002 replied:
PTAL again. Thanks

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+class QuestionSubmitterTotalContributionStatsModel(base_models.BaseModel):
+    """Records the question contribution stats for contributor dashboard admin
+    page. There is one instance of this model per contributor_id.
+    Its IDs is same as [contributor_id].
```
`The ID of each model is simply: [contributor_id].`

(There is only one ID for each model, so "IDs" is misleading.)

kshitij01042002 replied:
Okay. Changed.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+            page_size: int. Number of models to fetch.
+            offset: int. Number of results to skip from the beginning of all
+                results matching the query.
+            sort_by: str|None. A string indicating how to sort the result.
```
Same concerns as above.

kshitij01042002 replied:
Updated.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+            cls.get_all().filter(cls.contributor_id == user_id).fetch())
+        for model in stats_models:
+            splitted_id = model.id.split('.')
+            id_without_user_id = splitted_id[0]
```
I'm a bit confused by this. Why don't you just call it the language_code? The way it is written it sounds like some kind of ID.

kshitij01042002 replied:
Yeah, we could call it language code. Updated

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+    def get_model_association_to_user(
+    ) -> base_models.MODEL_ASSOCIATION_TO_USER:
+        """Model is exported as multiple instances per user since there are
+        multiple languages and topics relevant to a user.
```
Why are you referring to topics here?

kshitij01042002 replied:
Missed while editing. Updaed.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+
+    @classmethod
+    def get_deletion_policy(cls) -> base_models.DELETION_POLICY:
+        """Model contains corresponding to a user: contributor_id."""
```
This sentence doesn't scan.

kshitij01042002 replied:
It was actually supposed to be "Model contains data corresponding to a user: contributor_id.". Updated

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+            cls.get_all().filter(cls.contributor_id == user_id).fetch())
+        for model in stats_models:
+            splitted_id = model.id.split('.')
+            id_without_user_id = splitted_id[0]
```
Same question as above.

kshitij01042002 replied:
Updated.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+    Its IDs is same as [contributor_id].
+    """
+
+    # We use the model id as a key in the Takeout dict.
```
Why do we need to do this? We only have one model for each user.

kshitij01042002 replied:
I assumed we do this for every model. Changed it.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+    recent_performance = datastore_services.IntegerProperty(
+        required=True, indexed=True)
+    # Overall accuracy of the user.
+    # overall_accuracy = accepted questions/ submitted questions.
```
Add space before '/'. Similarly elsewhere.

kshitij01042002 replied:
Done.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+class QuestionReviewerTotalContributionStatsModel(base_models.BaseModel):
+    """Records the question review stats for contributor admin dashboard.
+    There is one instance of this model per contributor_id.
+    Its IDs is same as the user_id of the contributor.
```
Same comments as above, throughout this class.

kshitij01042002 replied:
Updated best upto my knowledge. PTAL. Thanks

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2000,1275 @@ def export_data(
+        for model in stats_models:
+            user_data = {
+                'topic_ids_with_question_reviews': (
+                model.topic_ids_with_question_reviews),
```
Indent by 4.

kshitij01042002 replied:
Done.

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -2994,3 +2994,1639 @@ def test_get_model_association_to_user(self) -> None:
+        sorted_results, next_offset, more = (
+            suggestion_models.TranslationSubmitterTotalContributionStatsModel
+            .fetch_page(
+            page_size=2,
```
All these args should be indented by 4.

Ditto below, in various places.

kshitij01042002 replied:
Updated.

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -2994,3 +2994,1639 @@ def test_get_model_association_to_user(self) -> None:
+            last_activity=None,
+            language_code='es'
+            ))
+        self.assertEqual(sorted_results[0].id, 'model_3')
```
Check len(sorted_results) and check sorted_results[1].id as well if it exists.

Ditto for other fetch_page() queries, do the test fully.

You should also be creating the data so that each sort produces a different result. You might need to check the second page of results as well for full confirmation. It is OK to make your page size larger when testing sort order, and have a separate test for checking pagination.

kshitij01042002 replied:
Updated tests. PTAL again. Thanks.

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -2994,3 +2994,1639 @@ def test_get_model_association_to_user(self) -> None:
+            page_size=1,
+            offset=0,
+            sort_by=constants.CD_ADMIN_STATS_SORT_OPTIONS[
+                    'DecreasingAccuracy'],
```
This should be 4 spaces indented from the previous line.

kshitij01042002 replied:
Updated

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -2994,3 +2994,1639 @@ def test_get_model_association_to_user(self) -> None:
+        self.assertTrue(more)
+        self.assertEqual(next_offset, 2)
+
+        # Check for last activity under 7 days.
```
I suggest creating a separate test for filters since they are conceptually different from sorts.

kshitij01042002 replied:
I have covered and tested every sort and filter we have implemented. Also every branch in the code. I am little confused here wdym by a separete test, do you mean we should create a different test class?

seanlip replied:
Similar to the other comment I left, I would suggest having different test methods (not classes) -- one for testing sorting, one for testing filtering, and one for testing the combo. Also, name your test methods clearly -- not just "test_method_name". See https://github.com/oppia/oppia/wiki/Backend-tests#guidelines-for-writing-good-tests for guidance.

It's generally not a good idea to have N different behavioural tests in a single method. If some issue causes multiple tests in that method to fail then it takes multiple test runs to root out all the errors. With different test methods you end up having better isolation and can see all the failures at once.

kshitij01042002 replied:
I have splitted tests into four tests. 
sorting
filtering
sorting and filtering
pagination.

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -2994,3 +2994,1639 @@ def test_get_model_association_to_user(self) -> None:
+            first_contribution_date=self.FIRST_CONTRIBUTION_DATE,
+            last_contribution_date=self.LAST_CONTRIBUTION_DATE
+        )
+        splitted_id_1 = model_1_id.split('.')
```
You shouldn't have logic in tests. Just hardcode the expected data.

kshitij01042002 replied:
Okay, changed but had doubt why shouldn't we have logic in tests and rather hardcode things only?

seanlip replied:
There are a lot of online posts on this (search for "avoid logic in tests"). Here's a random example http://www.everydayunittesting.com/2016/08/unit-test-anti-patterns-logic-in-tests.html but you can find many more with a web search.

kshitij01042002 replied:
Okay. thanks.

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -2994,3 +2994,1639 @@ def test_get_model_association_to_user(self) -> None:
+        self.assertTrue(more)
+        self.assertEqual(next_offset, 1)
+
+        # Check for language filter.
```
Same notes as before re splitting filtering to a separate test. Also note that last_activity is misspelled below.

Ditto for the other two models.

kshitij01042002 replied:
I think, the current test cases covers both sorting and filtering and we need to ensure that both works coherently as well. Separating the tests won't ensure that sorting and filtering works fine even when applied together. Also the same question as above by splitting you mean to split in two different test classes?

seanlip replied:
I would suggest having different test methods (not classes) -- one for testing sorting, one for testing filtering, and one for testing the combo. Also, name your test methods clearly -- not just "test_method_name". See https://github.com/oppia/oppia/wiki/Backend-tests#guidelines-for-writing-good-tests for guidance.

kshitij01042002 replied:
PTAL at new tests.

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -2994,3 +2994,1639 @@ def test_get_model_association_to_user(self) -> None:
+            first_contribution_date=self.FIRST_CONTRIBUTION_DATE,
+            last_contribution_date=self.LAST_CONTRIBUTION_DATE
+        )
+        splitted_id_1 = model_1_id.split('.')
```
Same notes as before re not doing logic in tests. Ditto for the other two models.

kshitij01042002 replied:
Changed.

kshitij01042002 committed:
first pass on reviews

kshitij01042002 committed:
modified tests

kshitij01042002 committed:
addressed all reviewers comments

kshitij01042002 committed:
fix lint errors

kshitij01042002 committed:
remaining review changes

seanlip commented on assets/constants.ts:
```diff
@@ -6993,6 +6993,6 @@ export default {
   "CONTRIBUTOR_CERTIFICATE_HEIGHT": 1313,
-  "BRANCH_NAME": "",
-  "SHORT_COMMIT_HASH": ""
+  "BRANCH_NAME": "gsoc_M1.2",
```
Woah. Why is this making it into the PR?

Please always check your "Files Changed" tab before requesting any reviews, in order to make sure the changes you submit are the ones you actually want to get merged. This is good PR discipline.

kshitij01042002 replied:
Sorry, this happened previously as well. I'll just remove it but any idea how this is automatically getting attached?

seanlip replied:
Try checking with @kevintab95 -- per https://github.com/oppia/oppia/blame/develop/assets/constants.ts#L7044 he merged that change 5 days ago.

kevintab95 replied:
Hi @kshitij01042002, few questions that will help me debug this:
1. When these files were committed, was the local server still active in the background?
2. If not, what method did you use to shut down the server?

Thanks!

kshitij01042002 replied:
1. No the local server was not running while commiting these changes.
2. I usually shutdown the server with ctrl+c, so did this time.

FYI these changes occured when I did a commit on my local branch.
Thanks @kevintab95.

kshitij01042002 replied:
Hi @kevintab95 these changes always takes place whenever I push changes to github

kevintab95 replied:
Hi @kshitij01042002, I tried this out but haven't been able to reproduce on my machine. The file constants.ts seems to be reset to the original state after I shutdown the server using the ctrl + c method. The changes however do show up if the server was not shutdown. Could you share the logs on the console corresponding to the shutdown process? This might help us understand what is happening.

kevintab95 replied:
Another thing I would suggest trying is, start the server in prod mode (i.e. use the "--prod_env" flag), wait for it to open up the splash page and then stop the server. Execute "git status" and see if there are any changes to the "DEV_MODE" constant in constants.ts -- prod mode also involves modifying the constants file temporarily. If there are changes, it probably means there is some issue with shutting down the server and we will need to investigate that.

kevintab95 replied:
Also, a note for the future, after you have responded to comment(s) and would like someone to take a look at it, please remember to assign them to the PR to avoid delays. The assignee field needs to reflect the individuals who need to take action on the PR. I almost missed this because I wasn't assigned.

kshitij01042002 replied:
Hi @kevintab95 I tried to run the server in prod mode and was exited with an error:-
![image](https://github.com/oppia/oppia/assets/91470808/1a00ba74-6176-4315-9013-388042ee07b9)

![image](https://github.com/oppia/oppia/assets/91470808/5cb6a6df-8f5e-4813-9ffa-81c5b380cbe3)

Also after doing this there were changes in index.yaml file and constants.ts file which wasn't reverted.

When I start the server in dev mode, there were changes in constants.ts but on shutting down the server the changes were reverted.

kevintab95 replied:
What changes were in the constants file when the server stopped -- when run in prod mode?

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -2004,9 +2025,9 @@ def export_data(
+    recent_review keyed per (language_code, contributor_id) tuple.
     Its IDs will be in the following
-        structure: [contributor_id][language_code]
+        structure: [language_code].[contributor_id]
```
Nit: Why the weird line break and indentation? (You can put "structure:" on the previous line and then have the rest on the next line.)

kshitij01042002 replied:
Actually we were exceeding line lenght therefore I did line break before structure.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -2137,7 +2192,11 @@ def get( # type: ignore[override]
     ) -> Optional[TranslationSubmitterTotalContributionStatsModel]:
         """Gets the TranslationSubmitterTotalContributionStatsModel
-        matching the supplied language_code, contributor_id, topic_id.
+        matching the supplied language_code, contributor_id.
```
...language_code and contributor_id.

kshitij01042002 replied:
Fixed.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -129,7 +133,24 @@
+NUM_MODELS_PER_FETCH: Final = 500
+
+
+class SortChoice():
```
Look at how other enums are done in the codebase, e.g. classes subclassing enum.Enum. Maybe you could use the same pattern.

Also the class name should be plural.

kshitij01042002 replied:
Changed

seanlip reviewed (changes requested):
I have to go afk, but did a partial review, will continue later.

@kshitij01042002 next time you want a full pass please make sure to update the list of assignees, thanks -- otherwise I might not see it.

kshitij01042002 committed:
covering backend test coverage

kshitij01042002 commented:
Hi @seanlip I have answered all your comments. PTAL. Thanks for the review.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.2

kshitij01042002 commented:
> I have to go afk, but did a partial review, will continue later.
> 
> @kshitij01042002 next time you want a full pass please make sure to update the list of assignees, thanks -- otherwise I might not see it.

Thanks @seanlip actually I was still addressing the comments therefore didn't asked for the full review. Now you can take a full review.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -121,6 +133,26 @@
+
+
+class SortChoices(enum.Enum):
+    """Enum for Sort Options available in Contributor Admin Dashboard"""
```
End docstrings with a period, just like elsewhere in the codebase.

kshitij01042002 replied:
Done. Also added some extra information for a new developer who might work on this in future,

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2022,1459 @@ def export_data(
+            sort_by: SortChoices|None. A string indicating how to sort the
+                result.
+            topic_ids: List[str]|None. List of topic ID(s) to fetch
+                contributor stats.
```
stats --> stats for.

kshitij01042002 replied:
Changed.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2022,1459 @@ def export_data(
+        """
+
+        sort_options_dict = {
+            constants.CD_ADMIN_STATS_SORT_OPTIONS[
```
I keep saying this again and again, and it doesn't seem to get addressed. To quote a previous review:

> Also, as mentioned in a previous comment, the mapping should be directly from SORT_KEY_INCREASING_LAST_ACTIVITY and you should be passing the sort keys to this function, rather than the human-readable values.

kshitij01042002 replied:
I am confused here about how you want flow to be. If you can see on line 2287 in am passing the sort keys to this function. Then it is checking on the basis of sort_key which sort query is appropriate.

seanlip replied:
If you're not sure about something, please feel free to ask for clarification rather than ignoring the comment.

What I am saying is that this dict should be:

```
sort_options_dict = {
     SortChoices.SORT_KEY_INCREASING_LAST_ACTIVITY.value: -cls.last_contribution_date,
     SortChoices.SORT_KEY_DECREASING_LAST_ACTIVITY.value: cls.last_contribution_date,
     ...
}
```

rather than what you have. Make sure that the input args and tests match this understanding as well. There should be no need to reference constants.ts.

kshitij01042002 replied:
So do we need to define values in constants.ts, should I remove those values? Or we might need them while building frontend?

kshitij01042002 replied:
Thanks @seanlip PTAL at the new changes if that looks good? Fow now I have kept the changes in the constants file as well. Thinking we might require it in future while building frontend.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2022,1459 @@ def export_data(
+                    after this batch.
+        """
+
+        sort_options_dict = {
```
Same notes as before, for this and the comment below. Ditto for other classes.

kshitij01042002 replied:
PTAL again. Thanks

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2022,1459 @@ def export_data(
+        # The first sort property must be the same as the property to which
+        # an inequality filter is applied. Thus, the inequality filter on
+        # last_activity can not be used here and we have implemented it below.
+        # Learn more about this here
```
Add a colon after "here" (ditto elsewhere).

kshitij01042002 replied:
Okay.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -1990,3 +2022,1459 @@ def export_data(
+
+    @classmethod
+    def get_deletion_policy(cls) -> base_models.DELETION_POLICY:
+        """Model contains corresponding to a user: contributor_id."""
```
This sentence doesn't scan, and I commented on this before.

When addressing a comment, please check other cases which might need to be fixed similarly; don't expect the reviewer to point all those cases out for you. (They often will, to an extent, but you should be doing the final audit.)

kshitij01042002 replied:
Changed this as well as other docstrings in this file to be accurate, Thanks.

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -2994,3 +2994,3029 @@ def test_get_model_association_to_user(self) -> None:
+                page_size=2,
+                offset=0,
+                sort_by=constants.CD_ADMIN_STATS_SORT_OPTIONS[
+                            'DecreasingLastActivity'],
```
As mentioned in a previous review, this should be 4 spaces indented from the previous line. It is not.

Ditto elsewhere in several other places.

kshitij01042002 replied:
I indented it after "=". Thanks for suggestio. Changed it.

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -2994,3 +2994,3029 @@ def test_get_model_association_to_user(self) -> None:
+        self.assertFalse(more)
+        self.assertEqual(next_offset, 2)
+
+    def test_fetch_page_for_pagination(self) -> None:
```
for --> with. Ditto for the others.

kshitij01042002 replied:
Changed.

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -2994,3 +2994,3029 @@ def test_get_model_association_to_user(self) -> None:
+                self.FIRST_CONTRIBUTION_DATE.isoformat()),
+            'last_contribution_date': (
+                self.LAST_CONTRIBUTION_DATE.isoformat())
+            }
```
Deindent this by 4 to match the opening {

kshitij01042002 replied:
Okay.

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -2994,3 +2994,3029 @@ def test_get_model_association_to_user(self) -> None:
+            last_contribution_date=self.LAST_CONTRIBUTION_DATE
+        )
+        expected_data = {
+                'topic_ids_with_question_submissions': (
```
This should be indented by 4 from the previous line.

In general your indentation in this PR is somewhat "off" -- there seem to be a number of such errors. Please double-check all of it.

seanlip reviewed (changes requested):
Thanks -- this is looking better! Left some more comments.

seanlip commented:
Please also see https://github.com/oppia/oppia/pull/18340#discussion_r1225424565, which is important.

kshitij01042002 committed:
added backend tests for enum and constants

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
docstring corrections

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
Merge branch 'gsoc_M1.2' of https://github.com/kshitij01042002/oppia into gsoc_M1.2

kshitij01042002 commented:
Hi @seanlip I have addressed your comments PTAL. Thanks

kshitij01042002 committed:
reverted wrong changes

kshitij01042002 committed:
removed branch name from constants.ts

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -6020,3 +6020,15 @@ def test_get_model_association_to_user(self) -> None:
+        for enum_member in suggestion_models.SortChoices:
+            self.assertTrue(
+                enum_member.value in constants.CD_ADMIN_STATS_SORT_OPTIONS
+            )
```
Check the other way round, too -- anything in that constants file must be in the enum.

I suggest getting the keys of each, sorting them, and making sure those lists are equal.

seanlip reviewed (changes requested):
This all looks good! Just one comment about the test. Thanks @kshitij01042002!

kshitij01042002 committed:
backend test for dict and enum match

kshitij01042002 commented:
Hi @seanlip added the test you suggested PTAL. Thanks for the review.

kshitij01042002 commented:
Hi @aasiffaizal can you also PTAL since Sean is almost finished with his review. Thanks.

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -6028,7 +6028,18 @@ class SortChoicesUnitTest(test_utils.GenericTestBase):
+        dict_values = []
         for enum_member in suggestion_models.SortChoices:
+            enum_values.append(enum_member.value)
             self.assertTrue(
```
You don't need this any more, right?

kshitij01042002 replied:
Yeah. it doesn't make sense to check this now.

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -6028,7 +6028,18 @@ class SortChoicesUnitTest(test_utils.GenericTestBase):
                 enum_member.value in constants.CD_ADMIN_STATS_SORT_OPTIONS
             )
+
+        for key in constants.CD_ADMIN_STATS_SORT_OPTIONS:
```
Just do `dict_values = constants.CD_ADMIN_STATS_SORT_OPTIONS.keys()` at the outset.

kshitij01042002 replied:
If we do this, we can't use the sort() function. I think we could keep the existing method as well.

seanlip replied:
I don't understand, why can't you use the sort() function? Don't you end up with a list of strings either way (perhaps after casting with list())?

kshitij01042002 replied:
Casting it into a list works. Thanks.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.2

kshitij01042002 committed:
removed obsolete test

kshitij01042002 committed:
lint errors

kshitij01042002 committed:
changed a test logic

kshitij01042002 committed:
removed branch name

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.2

kshitij01042002 commented:
Hi @seanlip are there any more changes to be made in the PR? I need this PR to be merged so that I can start working on the subsequent milestone that needs this PR merged.

Thanks.

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -2994,3 +2994,3049 @@ def test_get_model_association_to_user(self) -> None:
+        for enum_member in suggestion_models.SortChoices:
+            enum_values.append(enum_member.value)
+
+        dict_values = dict_values = list(
```
Why do you need dict_values = dict_values here?

kshitij01042002 replied:
Ahhh. So sorry for that. I'll just change that.

seanlip reviewed (changes requested):
Er, only one thing that could have been addressed by proofreading. Otherwise LGTM.

kshitij01042002 committed:
removed duplicate lines

kshitij01042002 committed:
Merge branch 'gsoc_M1.2' of https://github.com/kshitij01042002/oppia into gsoc_M1.2

seanlip reviewed (approved):
LGTM!

kshitij01042002 committed:
generated indexes in index.yaml

aasiffaizal reviewed (approved):
LGTM
